pkg/storagespace/storagespace.go

```diff
@@ -0,0 +1,147 @@
+)
+
+// SplitID splits a storage space ID into a provider ID and a node ID.
+// The accepted formats results of the storage space ID and respective results
+// are:
+// <providerid>$<spaceid>!<nodeid> 	-> <providerid>$<spaceid>, <nodeid>
+// <providerid>$<spaceid>			-> <providerid>$<spaceid>, <spaceid>
+// <spaceid>						-> <spaceid>, <spaceid>
+func SplitID(ssid string) (storageid, nodeid string, err error) {
+	if ssid == "" {
+		return "", "", errors.Wrap(ErrInvalidSpaceID, "can't split empty storage space ID")
+	}
+	parts := strings.SplitN(ssid, _idDelimiter, 2)
+	if len(parts) == 1 || parts[1] == "" {
+		_, sid := SplitStorageID(parts[0])
+		return parts[0], sid, nil
+	}
+	return parts[0], parts[1], nil
+}
+
+// SplitStorageID splits a storage ID into the provider ID and the spaceID.
+// The accepted formats are:
+// <providerid>$<spaceid>			-> <providerid>, <spaceid>
+// <spaceid>						-> "", <spaceid>
+func SplitStorageID(sid string) (providerID, spaceID string) {
+	parts := strings.SplitN(sid, _storageIDDelimiter, 2)
+	if len(parts) == 2 {
+		return parts[0], parts[1]
+	}
+	return "", parts[0]
+}
+
+// FormatResourceID converts a ResourceId into the string format.
+// The result format will look like:
+// <storageid>!<opaqueid>
+func FormatResourceID(sid provider.ResourceId) string {
+	return strings.Join([]string{sid.StorageId, sid.OpaqueId}, _idDelimiter)
+}
+
+// FormatStorageID converts the provider ID and space ID into the string format.
+// The result format will look like:
+// <providerid>$<spaceid>
+func FormatStorageID(providerID, spaceID string) string {
+	return strings.Join([]string{providerID, spaceID}, _storageIDDelimiter)
+}
+
+// ParseID parses a storage space ID and returns a storageprovider ResourceId.
+// The accepted formats are:
+// <providerid>$<spaceid>!<nodeid> 	-> <providerid>$<spaceid>, <nodeid>
+// <providerid>$<spaceid>			-> <providerid>$<spaceid>, <spaceid>
+// <spaceid>						-> <spaceid>, <spaceid>
+func ParseID(ssid string) (provider.ResourceId, error) {
+	sid, nid, err := SplitID(ssid)
+	return provider.ResourceId{
+		StorageId: sid,
+		OpaqueId:  nid,
+	}, err
+}
+
+// ParseReference parses a string into a spaces reference.
+// The expected format is `<providerid>$<spaceid>!<nodeid>/<path>`.
+func ParseReference(sRef string) (provider.Reference, error) {
+	parts := strings.SplitN(sRef, "/", 2)
+
+	rid, err := ParseID(parts[0])
+	if err != nil {
+		return provider.Reference{}, err
+	}
+
+	var path string
+	if len(parts) == 2 {
+		path = parts[1]
+	}
+
+	return provider.Reference{
+		ResourceId: &rid,
+		Path:       utils.MakeRelativePath(path),
+	}, nil
+}
+
+// FormatReference will format a storage space reference into a string representation.
+// If ref or ref.ResourceId are nil an error will be returned.
+// The function doesn't check if all values are set.
+// The resulting format can be:
+//
+// "storage_id!opaque_id"
+// "storage_id!opaque_id/path"
+// "storage_id/path"
```

Comment:
the `opaque_id` might contain `/`, eg for the localfs, which uses the path ins the fileid. we may have to change localfs to use another form of id ... it already uses sqlite. might as well use an identifier for fileid to path lookup.

I currently only see ̀ FormatReference` used in the share cache ... if so make this code a private function there. I fear people might expect to be able to parse `FormatReference` back, which might not always be the case.

Reply:
I think we have more places in the code where we manually build a storage space reference. Usually like `path.Join(storageid + opaqueid, relPath)`. But I think it would be beneficial to have that logic in a function so that changing the format of the reference string is easy. That is what this function should be for. If we don't need/want it fine. :shrug:

Reply:
Ok, that makes sense. Maybe add this intention to the doc string?

Reply:
> Maybe add this intention to the doc string?

I don't think we need a comment saying "Use this instead of building your own." Or do you mean something else?